### ISSUE TITLE
Validating CSS nesting not working

### ISSUE THREAD

**Issue by aaronbeall** [stylelint/stylelint#2269]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Trying to use rules to help convert a SCSS nested codebase (`postcss-nested`) to CSS standard nesting (`postcss-nesting`)

> Which rule, if any, is this issue related to?

`selector-nested-pattern`

> What CSS is needed to reproduce this issue?

```css
.this-is-wrong {
  .this-should-raise-error {
    color: red;
    .this-is-also-wrong {
      color: darkred;
    }
  }
}
.this-is-right {
  & .this-is-okay {
    color: green;
    & this-is-also-okay {
      color: darkgreen;
    }
  }
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "selector-nested-pattern": "^&:\\(\\?:hover\\|focus\\)$",
    "rule-nested-empty-line-before": [ "always", {
      "except": ["first-nested"],
      "ignore": ["after-comment"]
    } ]
  }
}
```

> Which version of stylelint are you using?

e.g. `7.7.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Webpack PostCSS plugin:

```js
  postcss: [
    require('postcss-import')({
      addDependencyTo: webpack,
      plugins: [
        require("stylelint")({ /* your options */ })
      ]
    }),
    require('postcss-cssnext'),
    require("postcss-reporter")({ clearMessages: true })
  ]
```

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

It relates to CSS spec nesting via `postcss-cssnext`.

> What did you expect to happen?

Warning or error flagged for the invalid nested rules, such as `.this-should-raise-error`, etc.

> What actually happened (e.g. what warnings or errors you are getting)?

No warnings or errors were raised.

**Comment by jeddy3** [comment]
@aaronbeall Thanks for using the template.

I think your regex might need some tweaking as the following config:

```json
{
  "rules": {
    "selector-nested-pattern": "^&"
  }
}
```

With this CSS:

```css
.this-is-wrong {
  .this-should-raise-error {
    color: red;
    .this-is-also-wrong {
      color: darkred;
    }
  }
}
.this-is-right {
  & .this-is-okay {
    color: green;
    & this-is-also-okay {
      color: darkgreen;
    }
  }
}
```

Correctly produces:

```console
2:3 error Expected nested selector ".this-should-raise-error" to match specified pattern (selector-nested-pattern)
4:5 error Expected nested selector ".this-is-also-wrong" to match specified pattern (selector-nested-pattern)
```

> Trying to use rules to help convert a SCSS nested codebase (`postcss-nested`) to CSS standard nesting (`postcss-nesting`)

If your intent here is to always enforce the use of the _direct nesting selector_ (`&`) then ` "selector-nested-pattern": "^&"` should suffice.

You can then use a combination of the following to restrict selectors and pseudos:

- `selector-max-compound-selectors`
- `selector-max-specificity`
- `selector-pseudo-class-whitelist`
- ...

**Comment by aaronbeall** [comment]
For some reason I still do not get any warnings or errors with `"selector-nested-pattern": "^&"`. Stylelint _is_ working because if I set another rule to something different than my codebase, for example indentation to `4` (my code uses 2 spaces) then I get tons of warning output. Do I need to set some other rules for the nesting pattern to take effect?

**Comment by jeddy3** [comment]
> Do I need to set some other rules for the nesting pattern to take effect?

Nope. All stylelint rules are standalone. You can have one or 180 of them in a configuration object and it should still work.

> FYI I copied the original regex string from the documentation example but the escaped characters broke things, which is why I double-escaped them...

Oh yeah, I think that ought to be `"^&:(?:hover|focus)$"` if it's a string. I'll patch that in PR.

> For some reason I still do not get any warnings or errors with `"selector-nested-pattern": "^&"`. 

I'm still unable to replicate it using http://stylelint.io/demo/. Can you create repo with a reproducible test in it please?

**Comment by jeddy3** [comment]
Closing. @aaronbeall Feel free to reopen if you're able to create a repo with a reproducible test in.

**Comment by aaronbeall** [comment]
No problem that's fair, it must be something specific to my setup but I can't isolate what it is... In the meantime I refactored all my code manually and am not using style-lint but certainly it's on my to-do list. :)

**Comment by woble** [comment]
@jeddy3 

We also use `postcss-nesting` and started having similar problem today after nesting and using `:not()` or `:matches()` with multiple parameters. I narrowed the problem down to the `selector-max-compound-selectors` rule. Running `stylelint@7.9.0`.

Tested the following on http://stylelint.io/demo/

```css
.bar {
  /* Breaks */
  @nest
  .foo[lorem='ipsum']:matches(:hover, :focus) & {
    & .fez {
      color: red;
    }
  }
  
  /* Breaks */
  @nest
  .foo[lorem='ipsum']:matches(.hover, .focus) & {
    & .fez {
      color: red;
    }
  }
  
  /* Doesn't break */
  @nest
  .foo[lorem='ipsum']:matches(:hover) & {
    & .fez {
      color: red;
    }
  }
  
  /* Doesn't break */
  & .fez {
    @nest
    .foo[lorem='ipsum']:matches(:hover, :focus) & {
      color: red;
    }
  }

  /* Doesn't break */
  @nest
  .foo[lorem='ipsum']:matches(:hover, :focus) & .fez {
    color: red;
  }
  
  /* Doesn't break */
  @nest
  .foo[lorem='ipsum']:matches(.hover) & {
    & .fez {
      color: red;
    }
  }
}
```

```json
{
  "rules":
  {
    "selector-max-compound-selectors": 3
  }
}
```

**Comment by evilebottnawi** [comment]
@woble can you do PR with failed tests? :+1:

**Comment by jeddy3** [comment]
@woble Thanks for the report. Can you also pull this out into a new issue please and use the issue template. I suspect resolving this will require some investigation (as there's a lot going on here - the use of `@nest` and multiple parameters for pseudo classes) and having the issue broken out into our normal template will help us.

**Comment by woble** [comment]
@jeddy3 
Sure thing, will do. I also just did a PR for the test as @evilebottnawi suggested.
